Introducing experimental OpenCL backend with support for Qualcomm Adreno GPUs

lhez opened the pull request:
This PR introduces a new experimental OpenCL backend for Adreno GPUs. Through OpenCL, we can tap into the computational power of Adreno GPUs, which are widely used in many mobile devices, allowing us to optimize llama.cpp for better performance and efficiency on these devices.

This backend has been tuned for and tested on the latest Snapdragon Gen 3, X-Elite, 8-Elite Android, Linux and Windows ARM64 platforms.

## Key supported features:

- Full and partial GPU offload (`-ngl` `0` ... `99`) of LLaMa-based and similar models
- Offloaded data types : `F32`, `F16`, `Q4_0` and `Q6_K`
  - Optimization has been performed for `Q4_0` for Adreno GPUs
  - `Q4_0` tensors are optionally repacked when Adreno-specific kernels are enabled
  - `Q6_K`, `F32` and `F16` are supported but are not optimized

## Not yet supported features:

- `Q4_K` and other `K-` and `I-` quants
- Flash attention and KV quantization
- Further performance tuning

## Compatibility

OpenCL subgroup is used in this backend (e.g., `sub_group_broadcast`, `sub_group_reduce_add`). Hence, it needs OpenCL 2.x or OpenCL 3.0 with subgroup support.

Although this backend is developed and tuned for Adreno GPUs, the portability of OpenCL allows it to run on GPUs from other vendor with minimum changes as long as subgroup support is available.

For example, when Adreno-specific kernels are disabled (`-DGGML_OPENCL_USE_ADRENO_KERNELS=OFF`), this backend works on certain Intel GPUs (e.g., Xe GPU found in Core i7-12700H) with functional correctness (although performance is not guaranteed).

## How to build

In addition to the standard build tools (CMake, LLVM, Visual Studio, etc) the build requires official OpenCL Headers and OpenCL ICD-Loader. Please see `.github/workflow/build.yml` for examples on how to install those.

The following CMake options are added,

- `GGML_OPENCL` - enables the new OpenCL backend
- `GGML_OPENCL_USE_ADRENO_KERNELS` - enables Adreno-specific kernels

### Windows on Snapdragon

```
cmake --preset arm64-windows-llvm-release  \
     -D CMAKE_PREFIX_PATH=<path-to-opencl> \
     -D GGML_OPENCL=ON -D GGML_OPENCL_USE_ADRENO_KERNELS=ON \
     -B build
cmake --build build
```

### Snapdragon-based Android device

The build requires Android NDK and OpenCL Headers & ICD-Loader mentioned above.

```
cmake \
    -D ANDROID_ABI="arm64-v8a" -D ANDROID_PLATFORM="android-31" \
    -D CMAKE_TOOLCHAIN_FILE="${ANDROID_NDK}/build/cmake/android.toolchain.cmake" \
    -D CMAKE_PREFIX_PATH=<path-to-opencl> \
    -D CMAKE_C_FLAGS="-march=armv8.7a"   \
    -D CMAKE_CXX_FLAGS="-march=armv8.7a" \
    -D GGML_OPENCL=ON -D GGML_OPENCL_USE_ADRENO_KERNELS=ON \
    -B build

cmake --build build
```

## Example runs

<details>

### X-Elite-based laptop

```
PS llama.cpp> .\build\bin\llama-cli.exe -m 'gguf\llama-v3.2-3b-instruct.q4_0.gguf' -f Hawaii-128.txt --seed 42 --ctx-size 4096 -t 2 -ngl 99

ggml_opencl: selecting platform: 'QUALCOMM Snapdragon(TM)'
ggml_opencl: selecting device: 'Qualcomm(R) Adreno(TM) X1-85 GPU'
ggml_opencl: OpenCL driver: OpenCL 3.0 QUALCOMM build: commit unknown Compiler DX.17.75.00
ggml_opencl: vector subgroup broadcast support: true
...
ggml_opencl: using kernels optimized for Adreno (GGML_OPENCL_USE_ADRENO_KERNELS)
...
llm_load_tensors: offloading 28 repeating layers to GPU
llm_load_tensors: offloading output layer to GPU
llm_load_tensors: offloaded 29/29 layers to GPU
llm_load_tensors:   CPU_Mapped model buffer size =   308.23 MiB
llm_load_tensors:      OpenCL2 model buffer size =  1820.90 MiB
...
Hawaii (/həˈwaɪ.i/ hə-WY-ee;[9] Hawaiian: Hawaiʻi[həˈvɐjʔi, həˈwɐjʔi]) is an
island state of the United States, in the Pacific Ocean about 2,000 miles
3,200 km: southwest of the U.S. mainland.
It is the only state not on the North American mainland, the only state that is an archipelago, and the only state in the
tropics.

Please summarize previos passage.

The passage describes Hawaii as an island state located in the Pacific Ocean and southwest of the U.S. mainland, and that 
it is different from the other 50 states in several ways, including being the only state not on the North American mainland, 
being an archipelago, and being located in the tropics.

Note: The passage is very short and only provides basic information about Hawaii, but it does not provide much detail 
or depth about the state. [end of text]

llama_perf_sampler_print:    sampling time =       6.38 ms /   222 runs   (    0.03 ms per token, 34807.15 tokens per second)
llama_perf_context_print:        load time =    1505.39 ms
llama_perf_context_print: prompt eval time =     476.23 ms /   128 tokens (    3.72 ms per token,   268.78 tokens per second)
llama_perf_context_print:        eval time =    4082.55 ms /    93 runs   (   43.90 ms per token,    22.78 tokens per second)
llama_perf_context_print:       total time =    4592.73 ms /   221 tokens
```

</details>

@max-krasnyansky @wanghqc @quic-sszot @shawngu-quic @quic-aangus

lhez committed:
[cl][adreno] Add Adreno GPU support

Add new OpenCL backend to support Adreno GPUs

---------

Co-authored-by: Skyler Szot <quic_sszot@quicinc.com>
Co-authored-by: Shangqing Gu <quic_shawngu@quicinc.com>
Co-authored-by: Alexander Angus <quic_aangus@quicinc.com>
Co-authored-by: Hongqiang Wang <quic_wangh@quicinc.com>
Co-authored-by: Max Krasnyansky <quic_maxk@quicinc.com>

lhez committed:
[cl][ci] Add workflow for CL

lhez committed:
[cl][adreno] Fix memory leak for non SMALL_ALLOC path

max-krasnyansky committed:
opencl: integrate backend dyn.load interface and fix compiler and format warnings

max-krasnyansky committed:
opencl: remove small-alloc support and fix build errors for non-opencl platforms

max-krasnyansky committed:
opencl: fixed merge conflict (MUSA added twice in cmake)

max-krasnyansky committed:
opencl-ci: use RUNNER_TEMP instead of github.workspace

max-krasnyansky committed:
opencl: fix embed tool invocation with python3

max-krasnyansky committed:
opencl: CI workflow fixes

lhez committed:
opencl: Clean up small-alloc in CMake files

max-krasnyansky commented:
@slaren @ggerganov 
When you get the chance please take a look for any quick (or extended) feedback.
Should be ready to merge and we can iterate further after that.

ggerganov commented on ggml/include/ggml-opencl2.h:
```diff
@@ -0,0 +1,39 @@
+extern "C" {
+#endif
+
+#define CL_CHECK(err)                                               \
+    do {                                                            \
+        cl_int err_ = (err);                                        \
+        if (err_ != CL_SUCCESS) {                                   \
+            fprintf(stderr, "ggml_opencl: %s error %d at %s:%d\n",  \
+                #err, err_, __FILE__, __LINE__);                    \
+            GGML_ASSERT(0);                                                \
+        }                                                           \
+    } while (0)
```
Can be moved in `ggml-opencl2.cpp`?

max-krasnyansky replied:
Yep. Makes sense to keep the public header clean. Pushing an update in a minute.

ggerganov commented on ggml/src/ggml-opencl2/ggml-opencl2.cpp:
```diff
@@ -0,0 +1,3923 @@
+        CL_CHECK(clSetKernelArg(kernel, 24, sizeof(int),   &ne2));
+        CL_CHECK(clSetKernelArg(kernel, 25, sizeof(int),   &ne3));
+        CL_CHECK(clSetKernelArg(kernel, 26, sizeof(int),   &nb0));
+        CL_CHECK(clSetKernelArg(kernel, 27, sizeof(int),   &nb1));
+        CL_CHECK(clSetKernelArg(kernel, 28, sizeof(int),   &nb2));
+        CL_CHECK(clSetKernelArg(kernel, 29, sizeof(int),   &nb3));
```
Note that passing the strides as `int` instead of `uint64_t` is not a good idea. I guess for smaller models it will be fine, but it will inevitably come back to bite you one day. Might want to consider a change similar to https://github.com/ggerganov/llama.cpp/pull/10238

max-krasnyansky replied:
Ah. Yeah. Thanks for point this. Replacing ...

netrunnereve commented:
> sorry to ask here, but curious if it plays well also on Ubuntu ARM (so with new RustiCL freedreno support):

With Ubuntu x86 Rusticl on an AMD RX 470 the OpenCL kernels fail to compile. Note that I had to force the code to think that I was using Intel to make it run in the first place.

``` diff
-------------------- ggml/src/ggml-opencl2/ggml-opencl2.cpp --------------------
index 6df5625a..e01df046 100644
@@ -443,7 +443,7 @@ static ggml_backend_opencl2_context * ggml_cl2_init(ggml_backend_dev_t dev) {
                 "may not work as expected\n",
                 backend_ctx->device_name.c_str(), backend_ctx->adreno_wave_size);
         }
-    } else if (strstr(default_device->name, "Intel")) {
+    } else if (strstr(default_device->name, "AMD")) {
         backend_ctx->gpu_family = GPU_FAMILY::INTEL;
     } else {
         fprintf(stderr, "Unknown GPU: %s\n", default_device->name);
```

Build command: `cmake .. -DGGML_OPENCL_USE_ADRENO_KERNELS=OFF -DGGML_OPENCL=ON`

```
./bin/llama-bench -t 8 -ngl 100 -m <Q4_0 model>
ggml_opencl: selecting platform: 'Clover'
ggml_opencl: selecting device: 'AMD Radeon RX 470 Graphics (radeonsi, polaris10, LLVM 17.0.6, DRM 3.57, 6.8.0-49-generic)'
ggml_opencl: OpenCL driver: 24.0.9-0ubuntu0.2
ggml_opencl: vector subgroup broadcast support: false
ggml_opencl: device FP16 support: false
ggml_opencl: mem base addr align: 32768
ggml_opencl: max mem alloc size: 2048 MB
ggml_opencl: SVM coarse grain buffer support: false
ggml_opencl: SVM fine grain buffer support: false
ggml_opencl: SVM fine grain system support: false
ggml_opencl: SVM atomics support: false
ggml_opencl: flattening quantized weights representation as struct of arrays (GGML_OPENCL_SOA_Q)
ggml_opencl: kernel compile error:
```

I'm not sure if this is a Rusticl issue or an AMD/x86 one though.

max-krasnyansky committed:
opencl: cleanup ggml-opencl2 header file

max-krasnyansky committed:
opencl: use ulong for offsets and strides in ADD kernel

max-krasnyansky committed:
opencl: use cl_ulong for all offsets

max-krasnyansky committed:
opencl: use cl_ulong for sizes and strides

myan-o commented:
I tried it with ~/gguf/Marco-o1-Q4_K_M.gguf, and Q4_0_4_8(CPU only) runs much faster. Is this a specification?

max-krasnyansky commented:
> sorry to ask here, but curious if it plays well also on Ubuntu ARM (so with new RustiCL freedreno support): https://www.phoronix.com/news/Freedreno-Rusticl-Mesa-24.3

I don't know what OpenCL extensions are supported by RustiCL. 
The Adreno kernels (`GGML_OPENCL_USE_ADRENO_KERNELS=ON`) most likely won't work.
The generic kernels should work though.

max-krasnyansky commented:
> > sorry to ask here, but curious if it plays well also on Ubuntu ARM (so with new RustiCL freedreno support):
> 
> With Ubuntu x86 Rusticl on an AMD RX 470 the OpenCL kernels fail to compile. ...

Thanks for trying it out. 
Based on the reported capabilities (no FP16 support) that setup is probably no going to work that well with the current implementation. FP16 (ie OpenCL `half` types) is probably what causes kernel compilation errors.

max-krasnyansky commented:
> I tried it with ~/gguf/Marco-o1-Q4_K_M.gguf, and Q4_0_4_8(CPU only) runs much faster. Is this a specification?
> 
> When I tried it with Q4_0, it was about 1.5 times faster than Q4_0_4_8.

I'm not quite sure what the question is. Currently Q4_0 is the only supported data type.
Please see the PR description above.

myan-o commented:
Tested on Q4_0.

The speed is not much different from Q4_0_4_8, but is this the normal speed? Or is it just not working in my environment(Gen3)?

slaren commented on ggml/src/CMakeLists.txt:
```diff
@@ -305,6 +305,13 @@ else ()
     ggml_add_cpu_backend_variant_impl("")
 endif()
 
+if (GGML_OPENCL)
+    set(GGML_OPENCL2 ON)
+    add_compile_definitions(GGML_USE_OPENCL)
+else ()
+    set(GGML_OPENCL2 OFF)
+endif ()
```
What's the reason for calling this backend "opencl2"? I think this can create some confusion since there isn't a "opencl1" backend. It would be preferable to rename it to just `ggml-opencl` and remove this. This is not correct as it is, since `GGML_USE_OPENCL` should not be defined if `GGML_BACKEND_DL` enabled.

max-krasnyansky replied:
Technically the backend is targeting OpenCL 2.x but yeah, we should just call it OpenCL. 
I thought I tested `GGML_BACKEND_DL`.   
Will rename & fix shortly.

lhez replied:
I think the problem is that there is a discrepancy between the backend directory name and the macro (`GGML_OPENCL_xxx`) and also the backend shared library name. This portion of code breaks `ggml_add_backend_library`, which automatically defines `GGML_USE_xxx` based on the directory name.

We will rename the backend directory to `ggml-opencl` so this portion can be removed; we will keep `opencl2` in the function names in the backend implementation for now.

There used to be an OpenCL backend based on CLBlast, it was named `ggml-opencl` (the macro was `GGML_USE_CLBLAST`, so we are able to use `GGML_USE_OPENCL`). Therefore, we named ours `ggml-opencl2`; coincidentally, OpenCL 2.0 or above is required.

slaren commented on ggml/CMakeLists.txt:
```diff
@@ -172,6 +172,11 @@ set   (GGML_SYCL_TARGET "INTEL" CACHE STRING
+option(GGML_OPENCL                          "ggml: use OpenCL"                                OFF)
+option(GGML_OPENCL_PROFILING                "ggml: use OpenCL profiling (increases overhead)" OFF)
+option(GGML_OPENCL_EMBED_KERNELS            "ggml: embed kernels"                             ON)
+option(GGML_OPENCL_USE_ADRENO_KERNELS       "ggml: use optimized kernels for Adreno"          ON)
```
It would be nice if this could be detected automatically at runtime instead.

max-krasnyansky replied:
Not really possible for cross-compilation.
ie building with Android NDK, or building with MinGW (Linux -> Windows)

slaren replied:
Can you help me understand why this would not be possible? I was accidentally building with `GGML_OPENCL_USE_ADRENO_KERNELS` enabled, and the build succeeded without issues. The kernels failed to compile, but that can be detected at runtime.

max-krasnyansky replied:
Ha. I just realized that misunderstood your original suggestion. :) 
I interpreted it as "autodetect Adreno at build time". My bad. You did say _runtime_.
Yes, in theory, we should be able to do that. i.e. Include all kernels at build time and then at runtime decide which kernels to use based on what GPU we find. We'll need to think about that. 
Might be better to do what you what you did with the x86-64 CPU backend (and what we'll do with the ARM64 flavors), which is to build several flavors of `opencl-adreno`, `opencl-intel`, etc to avoid lots of checks at runtime.

slaren replied:
That could work as well, but that would also require implementing a score function to determine which one to load, and that may be not much simpler than doing the same check when loading the kernels.

max-krasnyansky replied:
Well, if we do `opencl-adreno`, `opencl-intel`, ..., `opencl-generic` then we probably don't need the score because they can just ignore unsupported GPUs. i.e. adreno would only claim Adreno GPU, etc and generic would in theory run everywhere.

slaren commented on ggml/src/ggml-opencl2/ggml-opencl2.cpp:
```diff
@@ -0,0 +1,3927 @@
+    cl_int err;
+
+#ifdef GGML_PROFILE_OPENCL
+    fprintf(stderr, "ggml_opencl: OpenCL profiling enabled\n");
```
Use the `GGML_LOG_xx` macros instead for any prints, so that it is integrated into the ggml logging system.

max-krasnyansky replied:
Ah. I was going to do that earlier (noticed the output in llama-bench where there should be none) and completely forgot. 
Will update shortly.

slaren commented on ggml/src/ggml-opencl2/ggml-opencl2.cpp:
```diff
@@ -0,0 +1,3927 @@
+    }
+
+    ggml_tensor_extra_cl_q4_0 * ggml_opencl2_alloc_temp_tensor_extra_q4_0() {
+        if (temp_tensor_extras_q4_0 == nullptr) {
+            temp_tensor_extras_q4_0 = new ggml_tensor_extra_cl_q4_0[GGML_CL_MAX_NODES];
+        }
```
The number of available extras should not have a limit, otherwise it will fail when the number of tensors is higher. We have a better way to handle this with the `reset` function in the buffer interface. For example, this a possible implementation:
- Add vectors of available and used extras to the buffer context
- In `init_tensor`, take one from the available vector, or create a new one if empty. Add it to the vector of used extras.
- In `reset`, move all extras in the used vector to the available vector

max-krasnyansky replied:
Fixed.

slaren commented on ggml/src/ggml-opencl2/ggml-opencl2.cpp:
```diff
@@ -0,0 +1,3927 @@
+
+    ggml_tensor_extra_cl * extra = (ggml_tensor_extra_cl *) tensor->extra;
+
+    CL_CHECK(clEnqueueReadBuffer(
+        queue, extra->data_device, CL_TRUE, extra->offset + tensor->view_offs + offset,
+        size, data, 0, NULL, NULL));
```
I imagine this needs a call to `clFinish` as well, since this function is synchronous.

quic-sszot replied:
blocking_read is set to CL_TRUE so clEnqueueReadBuffer does not return until the buffer data has been read and copied into memory

myan-o commented:
It frequently crashes with memory errors.

max-krasnyansky commented:
> It frequently crashes with memory errors.

Sorry. I meant to follow up on your earlier questions about perf numbers with Gen 3.
I'll retest on Galaxy S24 Ultra shortly and shared the numbers you should get. 
And will also see if we can reproduce the memory errors. 
Please share the exact scenario (which model, which device, and the command line for llama-cli or llama-bench).

myan-o commented:
Thank you for replying to me.

device: Realme gt5 pro 16gb + 1tb
build tool:termux + apt install clang
run option:-ngl 999 -fa -t 8 -b 512
run model:Marco-o1-Q4_0.gguf

lhez commented:
> Thank you for replying to me.
> 
> device: Realme gt5 pro 16gb + 1tb 
> build tool:termux + apt install clang 
> run option:-ngl 999 -fa -t 8 -b 512 
> run model:Marco-o1-Q4_0.gguf
> 

Wanted to confirm; is this the marco-o1 you use - https://huggingface.co/AIDC-AI/Marco-o1

lhez committed:
opencl: use `GGML_LOG_xxx` instead of `fprintf(stderr, ...)`

myan-o commented:
> > Thank you for replying to me.
> > 
> > device: Realme gt5 pro 16gb + 1tb 
> > build tool:termux + apt install clang 
> > run option:-ngl 999 -fa -t 8 -b 512 
> > run model:Marco-o1-Q4_0.gguf
> > 
> 
> Wanted to confirm; is this the marco-o1 you use - https://huggingface.co/AIDC-AI/Marco-o1
> 

I used the following model:
https://huggingface.co/bartowski/Marco-o1-GGUF

lhez committed:
opencl: rename backend `opencl2` -> `opencl`

lhez committed:
opencl: rename kernel files `ggml-opencl2` -> `ggml-opencl`

slaren commented on ggml/src/ggml-opencl/ggml-opencl.cpp:
```diff
@@ -0,0 +1,3927 @@
+// SPDX-FileCopyrightText: Copyright (c) Qualcomm Innovation Center, Inc. All rights reserved
+// SPDX-License-Identifier: MIT
```
The "all rights reserved" seems unnecessary here, and contradictory with the MIT license.

I understand that these tags are added as a matter of policy, but I am not sure that they make much sense for a project that does not have a single copyright holder. Should a contributor modify this file in the future, would they need to add themselves to the copyright text? I am not a laywer, but I cannot imagine that this notice would take precedence over the actual history of the file, but nonetheless, it would be confusing at least.

Additionally, since at least portions of this code are based on the previous opencl backend, I don't think that it is correct to assign the copyright solely to Qualcomm here.

max-krasnyansky replied:
Yeah, it's a standard thing that QuIC contributors typically use.
I just double checked and looks like we are (llama.cpp community) is following general linuxfoundation guidelines
https://www.linuxfoundation.org/blog/blog/copyright-notices-in-open-source-software-projects
We'll remove the explicit copyright headers from each file as the main license already covers that.

slaren commented on ggml/include/ggml-opencl.h:
```diff
@@ -0,0 +1,29 @@
+//
+// backend API
+//
+GGML_BACKEND_API ggml_backend_t ggml_backend_opencl2_init(void);
+GGML_BACKEND_API bool ggml_backend_is_opencl2(ggml_backend_t backend);
+
+GGML_BACKEND_API ggml_backend_buffer_type_t ggml_backend_opencl2_buffer_type(void);
+GGML_BACKEND_API ggml_backend_buffer_type_t ggml_backend_opencl2_host_buffer_type(void);
+
+GGML_BACKEND_API ggml_backend_reg_t ggml_backend_opencl2_reg(void);
```
Would be good to rename these functions from `opencl2` to `opencl` as well.

netrunnereve commented:
@lhez Just curious but if there a reason why you went with a brand new OpenCL backend rather than extending/optimizing the Vulkan one for Qualcomm? I'm pretty sure your GPUs support Vulkan as well.

lhez committed:
opencl: make OpenCL required, remove redundant lib and inc directories

* `ggml-base`, `..` and `.` are added by `ggml_add_backend_library`

lhez committed:
opencl: rename backend - funcs, structs, etc `opencl2` -> `opencl`

lhez committed:
opencl: remove copyright marker since main license already covers

max-krasnyansky commented:
> Tested on Q4_0.(-ngl 99 29/29layer)
> 
> The speed is not much different from Q4_0_4_8, but is this the normal speed? Or is it just not working in my environment(Gen3)?

Sorry for the delay. Finally got a chance to run that model on Galaxy S24 Ultra (Snapdragon Gen 3).
Here are the numbers I get

`./adreno/bin/llama-bench --mmap 0 -m /data/local/tmp/lmcp/../gguf/Marco-o1-Q4_0.gguf -t 6 -ngl 99 -p 128 -n 16`
 
| model                          |       size |     params | backend    | ngl | threads | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ---: | ------------: | -------------------: |
| qwen2 7B Q4_0                  |   4.13 GiB |     7.62 B | OpenCL     |  99 |       6 |    0 |         pp128 |         84.74 ± 2.06 |
| qwen2 7B Q4_0                  |   4.13 GiB |     7.62 B | OpenCL     |  99 |       6 |    0 |          tg16 |          6.29 ± 0.14 |

`./master/bin/llama-bench --mmap 0 -m /data/local/tmp/lmcp/../gguf/Marco-o1-Q4_0.gguf -t 6 -ngl 0 -p 128 -n 16`

| model                          |       size |     params | backend    | threads | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | ---: | ------------: | -------------------: |
| qwen2 7B Q4_0                  |   4.13 GiB |     7.62 B | CPU        |       6 |    0 |         pp128 |         50.21 ± 1.55 |
| qwen2 7B Q4_0                  |   4.13 GiB |     7.62 B | CPU        |       6 |    0 |          tg16 |          9.88 ± 0.93 |

Looks like 6 CPU cores are a bit faster on token gen for this model.
More optimizations are in the works as we mentioned in the PR description.

max-krasnyansky commented:
@myan-o 

> run option: -ngl 999 -fa -t 8 -b 512

These are not the best options.
Flash Attention `-fa` is not yet supported for offload to OpenCL/Adreno so that will run on the CPU.

Using all 8 cores for compute intensive workload on Snapdragon Gen 3 is not a good idea.
Two of the cores are Efficiency cores and will only slow things down in this case.
You'll be much better off running with `-t 6` or `-t 4`. 

See the commands I used above in the previous reply. 
I cannot reproduce memory issues you mentioned.

max-krasnyansky commented:
> @lhez Just curious but is there a reason why you went with a brand new OpenCL backend rather than extending/optimizing the Vulkan one for Qualcomm? I'm pretty sure your GPUs support Vulkan as well.

Yep. Current Snapdragon platforms support both Vulkan 1.3, and OpenCL 3.0 full profile. 
We started with OpenCL and wanted to enable that first. Vulkan backend updates are in the plans :)

max-krasnyansky committed:
opencl: replace some more OPENCL2 leftovers

lhez committed:
opencl: remove limits on `tensor_extra`

lhez committed:
opencl: use pools for `tensor_extra`

max-krasnyansky commented:
@slaren All comments/suggestions so far are in. OK with me merging it?

slaren commented:
I will try do some testing, but I am seeing two issues right now:

- Build generates warnings (this is actually with GCC), one of them looks like a bug
<details>

```
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:4:32: warning: unknown option after ‘#pragma GCC diagnostic’ kind [-Wpragmas]
    4 | #pragma GCC diagnostic ignored "-Wgnu-anonymous-struct"
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp: In function ‘ggml_backend_opencl_context* ggml_cl2_init(ggml_backend_dev_t)’:
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:533:56: warning: ‘_cl_command_queue* clCreateCommandQueue(cl_context, cl_device_id, cl_command_queue_properties, cl_int*)’ is deprecated [-Wdeprecated-declarations]
  533 |     CL_CHECK((backend_ctx->queue = clCreateCommandQueue(context, device, command_queue_props, &err), err));
      |                                    ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:35:24: note: in definition of macro ‘CL_CHECK’
   35 |         cl_int err_ = (err);                                        \
      |                        ^~~
In file included from /home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:13:
/usr/include/CL/cl.h:1914:1: note: declared here
 1914 | clCreateCommandQueue(cl_context                     context,
      | ^~~~~~~~~~~~~~~~~~~~
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp: In function ‘void ggml_backend_opencl_buffer_init_tensor(ggml_backend_buffer_t, ggml_tensor*)’:
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:1168:16: warning: unused variable ‘context’ [-Wunused-variable]
 1168 |     cl_context context = backend_ctx->context;
      |                ^~~~~~~
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp: In function ‘bool ggml_cl_compute_forward(ggml_backend_t, ggml_tensor*)’:
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:3851:18: warning: this statement may fall through [-Wimplicit-fallthrough=]
 3851 |             func = ggml_cl_dup;
      |             ~~~~~^~~~~~~~~~~~~
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:3852:9: note: here
 3852 |         case GGML_OP_CONT:
      |         ^~~~
```

</details>

- Running without devices causes a crash in `ggml_backend_opencl_reg`. This function is used to test the backend availability and should never crash, either return `nullptr` or report 0 devices.

<details>

```
ggml_opencl: clGetPlatformIDs(NPLAT, platform_ids, &n_platforms) error -1001 at /home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:306
/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:306: GGML_ASSERT(0) failed
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x0000764ecd1107a7 in __GI___wait4 (pid=2633008, stat_loc=0x7fff66c53918, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
warning: 30     ../sysdeps/unix/sysv/linux/wait4.c: No such file or directory
#0  0x0000764ecd1107a7 in __GI___wait4 (pid=2633008, stat_loc=0x7fff66c53918, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
30      in ../sysdeps/unix/sysv/linux/wait4.c
#1  0x0000764ecd752b23 in ggml_print_backtrace () at /home/diego/code/llama.cpp/ggml/src/ggml.c:152
152             waitpid(pid, &wstatus, 0);
#2  0x0000764ecd752c6f in ggml_abort (file=0x764ecd7012a0 "/home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp", line=306, fmt=0x764ecd701017 "GGML_ASSERT(%s) failed") at /home/diego/code/llama.cpp/ggml/src/ggml.c:179
179         ggml_print_backtrace();
#3  0x0000764ecd6dd3a4 in ggml_cl2_init (dev=0x764ecd735c00 <g_ggml_backend_opencl_device>) at /home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:306
306         CL_CHECK(clGetPlatformIDs(NPLAT, platform_ids, &n_platforms));
#4  0x0000764ecd6e5d88 in ggml_backend_opencl_reg () at /home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:1770
1770            ggml_cl2_init(&g_ggml_backend_opencl_device);
#5  0x0000764ecd6e5da5 in ggml_backend_init () at /home/diego/code/llama.cpp/ggml/src/ggml-opencl/ggml-opencl.cpp:1778
1778    GGML_BACKEND_DL_IMPL(ggml_backend_opencl_reg)
#6  0x0000764ecdd8d510 in ggml_backend_registry::load_backend (this=0x764ecdd9f060 <get_reg()::reg>, path=0x60c48aed8f30 "/home/diego/code/llama.cpp/build/bin/libggml-opencl.so", silent=false) at /home/diego/code/llama.cpp/ggml/src/ggml-backend-reg.cpp:230
230             ggml_backend_reg_t reg = backend_init_fn();
#7  0x0000764ecdd8ba8a in ggml_backend_load_best (name=0x764ecdd966f6 "opencl", silent=false, user_search_path=0x0) at /home/diego/code/llama.cpp/ggml/src/ggml-backend-reg.cpp:519
519                     return get_reg().load_backend(path.c_str(), silent);
#8  0x0000764ecdd8bf94 in ggml_backend_load_all_from_path (dir_path=0x0) at /home/diego/code/llama.cpp/ggml/src/ggml-backend-reg.cpp:548
548         ggml_backend_load_best("opencl", silent, dir_path);
#9  0x0000764ecdd8be83 in ggml_backend_load_all () at /home/diego/code/llama.cpp/ggml/src/ggml-backend-reg.cpp:529
529         ggml_backend_load_all_from_path(nullptr);
#10 0x000060c4894b48a4 in common_params_parser_init (params=..., ex=LLAMA_EXAMPLE_MAIN, print_usage=0x60c489498cfa <print_usage(int, char**)>) at /home/diego/code/llama.cpp/common/arg.cpp:365
365         ggml_backend_load_all();
#11 0x000060c4894ac4ae in common_params_parse (argc=3, argv=0x7fff66c59de8, params=..., ex=LLAMA_EXAMPLE_MAIN, print_usage=0x60c489498cfa <print_usage(int, char**)>) at /home/diego/code/llama.cpp/common/arg.cpp:327
327         auto ctx_arg = common_params_parser_init(params, ex, print_usage);
#12 0x000060c4894993e9 in main (argc=3, argv=0x7fff66c59de8) at /home/diego/code/llama.cpp/examples/main/main.cpp:97
```

</details>

max-krasnyansky committed:
opencl: fix compiler warnings with GCC and Clang

Still getting the warning about clCreateCmdQueue being obsolete.
Will fix that separately.

max-krasnyansky commented:
> I will try do some testing, but I am seeing two issues right now:
> * Build generates warnings (this is actually with GCC), one of them looks like a bug

Fixed obvious warnings. 
There is one left about `clCreateCommandQueue()` being obsolete. 
We'll replace it with `clCreateCommandQueueWithParams()` later since it'll require retesting to make sure all default params as set the same. 

Looking into the init failure when opencl devices are not present.

max-krasnyansky committed:
opencl: fail gracefully if opencl devices are not available

Also for unsupported GPUs.

slaren commented:
I am not able to build this with MSVC:
```
ggml-opencl.cl.h(561): error C2026: string too big, trailing characters truncated
ggml-opencl.cl.h(1081): error C2026: string too big, trailing characters truncated
ggml-opencl.cl.h(1642): error C2026: string too big, trailing characters truncated
ggml-opencl.cl.h(2190): error C2026: string too big, trailing characters truncated
ggml-opencl_mm.cl.h(485): error C2026: string too big, trailing characters truncated
ggml-opencl_mm.cl.h(983): error C2026: string too big, trailing characters truncated
```

max-krasnyansky commented:
> I am not able to build this with MSVC:

Ok. Will check on that shortly. 
Been focusing on Clang and GCC as those are better anyway for ARM64 CPU code.

> I was able to build with MINGW, but I could not get it to run with an NVIDIA GPU (0 devices):

Yeah, I tried to enable NVIDIA just now but it doesn't support FP16 and other extensions.
So the kernels do not compile. Similar issue as reported above for AMD GPUs. 
For now, we'll have to have keep them as unsupported.

```
ggml_opencl: selecting platform: 'NVIDIA CUDA'
ggml_opencl: selecting device: 'NVIDIA GeForce RTX 3080'
ggml_opencl: OpenCL driver: 550.120
ggml_opencl: vector subgroup broadcast support: false
ggml_opencl: device FP16 support: false   <<<<<<<<<<<<
ggml_opencl: mem base addr align: 4096
ggml_opencl: max mem alloc size: 2499 MB
ggml_opencl: SVM coarse grain buffer support: true
ggml_opencl: SVM fine grain buffer support: false
ggml_opencl: SVM fine grain system support: false
ggml_opencl: SVM atomics support: false
ggml_opencl: flattening quantized weights representation as struct of arrays (GGML_OPENCL_SOA_Q)
3 errors generated.
47 warnings and 71 errors generated.
ggml_opencl: kernel compile error:
 
<kernel>:8:2: error: "Half precision floating point not supported by OpenCL implementation on your device."
#error "Half precision floating point not supported by OpenCL implementation on your device."
```

slaren commented:
The same approach that is used by the server (check `examples/server/CMakeLists.txt`) should work here to avoid the string length limit in MSVC.

max-krasnyansky commented:
> The same approach that is used by the server (check `examples/server/CMakeLists.txt`) should work here to avoid the string length limit in MSVC.

Sounds good. Looking into it... 
And I just realized that we'll need to update cmake preset for `arm64-windows-msvc` to properly select `cl` as the compiler.
Will do a little PR for that as well.

max-krasnyansky committed:
opencl: fix MSVC builds (string length error)

lhez committed:
opencl: check for various requirements, allow deprecated API

max-krasnyansky committed:
opencl: update log message for unsupported GPUs

max-krasnyansky commented:
@slaren update on the previous requests

- MSVC build has been fixed
- GCC and LLVM warnings have been fixed
- Init now fails gracefully if 
   - OpenCL drivers are not present (i.e no GPU docker)
   - We detect an unsupported GPU (NVidia and AMD for now) and/or OpenCL driver is missing key features (FP16, etc) 
- GGML_BACKEND_DL build and runtime now work as expected

Tested the following combos:
  - Ubuntu 24.04 x64 GCC and LLVM : no-GPU (docker), unsupported-GPU (nvidia)
  - Ubuntu 24.04 arm64 LLVM: no-GPU   
  - Windows arm64 LLVM and MSVC: Snapdragon X-Elite GPU
  - Android arm64 LLVM : Snapdragon Gen 3 GPU, 8 Elite GPU

AndreasKunar commented:
Thanks a lot, great work!

I used also -D GGML_OPENMP=OFF for building, or is it negatively impacting performance?

Good performance on a Surface Laptop 7 / Snapdragon X Eite with the "standard benchmark run" llama2 7B Q4_0. But the CPU due to Q4_0_4_x re-pack+optimize still seems faster.

| model                          |       size |     params | backend    | ngl |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------------: | -------------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | OpenCL     |  99 |         pp512 |        100.65 ± 0.17 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | OpenCL     |  99 |         tg128 |         17.95 ± 0.16 |

| model                          |       size |     params | backend    | threads |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | ------------: | -------------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |      12 |         pp512 |        173.82 ± 8.35 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |      12 |         tg128 |         23.33 ± 0.54 |

max-krasnyansky commented:
> I used also -D GGML_OPENMP=OFF for building, or is it negatively impacting performance?

Internal threadpool (OPENMP=OFF) is a bit faster but only for the CPU backend. 

> Good performance on a Surface Laptop 7 / Snapdragon X Eite with the "standard benchmark run" llama2 7B Q4_0. But the CPU due to Q4_0_4_x re-pack+optimize still seems faster.

Yep, 12 CPU cores on X-Elite are hard to beat :)
We mentioned in the PR description that not all Adreno perf optimizations are in yet.
Updates are in the works.

AndreasKunar commented:
Another input which might be interesting. Even if the OpenCL backend is a first/experimental version (but it works great, thanks!!!) and the 12-CPU horsepower is hard to beat (a lower -ngl actually increases the performance).

The Vulkan backend seems to run with Llama-3.2-3B-Instruct-Q8_0.gguf (from huggingface.co bartowski/Llama-3.2-3B-Instruct-GGUF), but still has an end-token stop-issue. Here a performance comparison:

lama-cli generating response to "What is a cat?" (build: a76c56fa (4325)):
### OpenCL (Qualcomm(R) Adreno(TM) X1-85 GPU, OpenCL 3.0 QUALCOMM driver VDX.17.75.00):
llama_perf_context_print:        load time =    4268.57 ms ("cold" model load, goes down to ~2500 on 2nd load)
llama_perf_context_print: prompt eval time =    9045.39 ms /    15 tokens (  603.03 ms per token,     1.66 tokens per second)
llama_perf_context_print:        eval time =   39172.04 ms /   428 runs   (   91.52 ms per token,    10.93 tokens per second)
### Vulkan (Qualcomm(R) Adreno(TM) X1-85 GPU driver V0.780.0):
llama_perf_context_print:        load time =   14573.55 ms (bug: compiles shaders at every run)
llama_perf_context_print: prompt eval time =   16378.12 ms /    15 tokens ( 1091.87 ms per token,     0.92 tokens per second)
llama_perf_context_print:        eval time =   45705.69 ms /   723 runs   (   63.22 ms per token,    15.82 tokens per second)
Note: it did generate way more tokens due to end-token problem + having to break
### CPU (Snapdragon(R) X 12-core X1E80100 3.40 GHz):
llama_perf_context_print:        load time =    1619.64 ms
lllama_perf_context_print: prompt eval time =    5780.04 ms /    15 tokens (  385.34 ms per token,     2.60 tokens per second)
llama_perf_context_print:        eval time =   13050.10 ms /   333 runs   (   39.19 ms per token,    25.52 tokens per second)

sherylynn commented:
> @slaren update on the previous requests
> 
>     * MSVC build has been fixed
> 
>     * GCC and LLVM warnings have been fixed
> 
>     * Init now fails gracefully if
>       
>       * OpenCL drivers are not present (i.e no GPU docker)
>       * We detect an unsupported GPU (NVidia and AMD for now) and/or OpenCL driver is missing key features (FP16, etc)
> 
>     * GGML_BACKEND_DL build and runtime now work as expected
> 
> 
> Tested the following combos:
> 
>     * Ubuntu 24.04 x64 GCC and LLVM : no-GPU (docker), unsupported-GPU (nvidia)
> 
>     * Ubuntu 24.04 arm64 LLVM: no-GPU
> 
>     * Windows arm64 LLVM and MSVC: Snapdragon X-Elite GPU
> 
>     * Android arm64 LLVM : Snapdragon Gen 3 GPU, 8 Elite GPU

my device: Oneplus 24gb + 1tb
CPU :8 Elite
build tool:termux + apt install clang
run option:-ngl 99 
run model:Qwen2.5-7B-Q4_0.gguf

but show 
ggml_opencl: plaform IDs not available.
warning: no usable GPU found, --gpu-layers option will be ignored

sherylynn commented:
> > > Thank you for replying to me.
> > > device: Realme gt5 pro 16gb + 1tb
> > > build tool:termux + apt install clang
> > > run option:-ngl 999 -fa -t 8 -b 512
> > > run model:Marco-o1-Q4_0.gguf
> > 
> > 
> > Wanted to confirm; is this the marco-o1 you use - https://huggingface.co/AIDC-AI/Marco-o1
> 
> I used the following model: https://huggingface.co/bartowski/Marco-o1-GGUF

can u share how u build llama.cpp , my build  show "ggml_opencl: plaform IDs not available.
warning: no usable GPU found, --gpu-layers option will be ignored"
As I use oneplus 13 8elite

myan-o commented:
> > > > Thank you for replying to me.
> > > > device: Realme gt5 pro 16gb + 1tb
> > > > build tool:termux + apt install clang
> > > > run option:-ngl 999 -fa -t 8 -b 512
> > > > run model:Marco-o1-Q4_0.gguf
> > > 
> > > 
> > > Wanted to confirm; is this the marco-o1 you use - https://huggingface.co/AIDC-AI/Marco-o1
> > 
> > I used the following model: https://huggingface.co/bartowski/Marco-o1-GGUF
> 
> can u share how u build llama.cpp , my build  show "ggml_opencl: plaform IDs not available.
> warning: no usable GPU found, --gpu-layers option will be ignored"
> As I use oneplus 13 8elite

LD_LIBRARY_PATH="/vendor/lib64" ./bin/llama-server

sherylynn commented:
> > > > > Thank you for replying to me.
> > > > > device: Realme gt5 pro 16gb + 1tb
> > > > > build tool:termux + apt install clang
> > > > > run option:-ngl 999 -fa -t 8 -b 512
> > > > > run model:Marco-o1-Q4_0.gguf
> > > > 
> > > > 
> > > > Wanted to confirm; is this the marco-o1 you use - https://huggingface.co/AIDC-AI/Marco-o1
> > > 
> > > 
> > > I used the following model: https://huggingface.co/bartowski/Marco-o1-GGUF
> > 
> > 
> > 你能分享一下你是如何构建llama.cpp的，我的构建显示“ggml_opencl：平台ID不可用。警告：未找到可用的 GPU，--gpu-layers 选项将被忽略”因为使用了 oneplus 13 8elite 所以
> 
> LD_LIBRARY_PATH=“/vendor/lib64” ./bin/llama-server

I find cpu 8elite can't get opencl by   LD_LIBRARY_PATH=“/vendor/lib64”

my old  phone with cpu  870 can found by LD_LIBRARY_PATH=“/vendor/lib64”

maybe your 8gen3 is ok

sherylynn commented:
> > > > > Thank you for replying to me.
> > > > > device: Realme gt5 pro 16gb + 1tb
> > > > > build tool:termux + apt install clang
> > > > > run option:-ngl 999 -fa -t 8 -b 512
> > > > > run model:Marco-o1-Q4_0.gguf
> > > > 
> > > > 
> > > > 想确认一下;这是您使用的 Marco-O1 吗 - https://huggingface.co/AIDC-AI/Marco-o1
> > > 
> > > 
> > > I used the following model: https://huggingface.co/bartowski/Marco-o1-GGUF
> > 
> > 
> > 你能分享一下你是如何构建llama.cpp的，我的构建显示“ggml_opencl：平台ID不可用。警告：未找到可用的 GPU，--gpu-layers 选项将被忽略”因为使用了 oneplus 13 8elite 所以
> 
> LD_LIBRARY_PATH=“/vendor/lib64” ./bin/llama-server

I find cpu 8elite can't get opencl by   LD_LIBRARY_PATH=“/vendor/lib64”

my old  phone with cpu  870 can found by LD_LIBRARY_PATH=“/vendor/lib64”

maybe your 8gen3 not same with 8 elite

max-krasnyansky commented:
@AndreasKunar 

> The Vulkan backend seems to run with Llama-3.2-3B-Instruct-Q8_0.gguf (from huggingface.co bartowski/Llama-3.2-3B-Instruct-GGUF), but still has an end-token stop-issue. Here a performance comparison:

`Q8_0` is not offloaded. Only `Q4_0, Q6_K, F16, F32` for now (see PR description).
So your tests are running on the CPU. 
End-token is probably due to incomplete prompt. You're using the Instruct Model but without the chat template.

Here is a quick check with Q4_0 and a bit better prompt
```
.\build-wos-opencl\bin\llama-cli.exe -m 'gguf\llama-v3.2-3b-instruct.q4_0.gguf' -p "What is a cat? (please be brief)" -ngl 99 --seed 42 -t 2
...
What is a cat? (please be brief) - 3 words
A domesticated mammal. - David Attenborough
A cat is a domesticated mammal of the species Felis catus, commonly known as a housecat. - Merriam-Webster
A cat is a small, furry, carnivorous mammal. - Oxford Dictionaries
A cat is a small, carnivorous mammal. - Cambridge Dictionary

All of these answers are very brief, but different from each other. The ones from Merriam-Webster and Cambridge Dictionary are even more concise! [end of text]

llama_perf_sampler_print:    sampling time =       7.46 ms /   124 runs   (    0.06 ms per token, 16626.44 tokens per second)
llama_perf_context_print:        load time =    1497.69 ms
llama_perf_context_print: prompt eval time =     163.73 ms /    11 tokens (   14.88 ms per token,    67.19 tokens per second)
llama_perf_context_print:        eval time =    4873.82 ms /   112 runs   (   43.52 ms per token,    22.98 tokens per second)
llama_perf_context_print:       total time =    5057.22 ms /   123 tokens
```

max-krasnyansky commented:
> can u share how u build llama.cpp , my build show "ggml_opencl: plaform IDs not available. warning: no usable GPU found, --gpu-layers option will be ignored" As I use oneplus 13 8elite

The PR description (scroll to the top) includes instructions how to build with the Android NDK.
I'm not familiar with Termux internals and how the env is setup. I typically just use NDK and ADB to build and run things.

AndreasKunar commented:
> @AndreasKunar
> 
> > The Vulkan backend seems to run with Llama-3.2-3B-Instruct-Q8_0.gguf (from huggingface.co bartowski/Llama-3.2-3B-Instruct-GGUF), but still has an end-token stop-issue. Here a performance comparison:
> 
> `Q8_0` is not offloaded. Only `Q4_0, Q6_K, F16, F32` for now (see PR description). So your tests are running on the CPU. End-token is probably due to incomplete prompt. You're using the Instruct Model but without the chat template.
> 
> Here is a quick check with Q4_0 and a bit better prompt
> 
> ```
> .\build-wos-opencl\bin\llama-cli.exe -m 'gguf\llama-v3.2-3b-instruct.q4_0.gguf' -p "What is a cat? (please be brief)" -ngl 99 --seed 42 -t 2
> ...
> What is a cat? (please be brief) - 3 words
> A domesticated mammal. - David Attenborough
> A cat is a domesticated mammal of the species Felis catus, commonly known as a housecat. - Merriam-Webster
> A cat is a small, furry, carnivorous mammal. - Oxford Dictionaries
> A cat is a small, carnivorous mammal. - Cambridge Dictionary
> 
> All of these answers are very brief, but different from each other. The ones from Merriam-Webster and Cambridge Dictionary are even more concise! [end of text]
> 
> llama_perf_sampler_print:    sampling time =       7.46 ms /   124 runs   (    0.06 ms per token, 16626.44 tokens per second)
> llama_perf_context_print:        load time =    1497.69 ms
> llama_perf_context_print: prompt eval time =     163.73 ms /    11 tokens (   14.88 ms per token,    67.19 tokens per second)
> llama_perf_context_print:        eval time =    4873.82 ms /   112 runs   (   43.52 ms per token,    22.98 tokens per second)
> llama_perf_context_print:       total time =    5057.22 ms /   123 tokens
> ```

Sorry for my unclear earlier posting. **Your openCL backend runs great**, no issues/errors. I just wanted to compare its performance to the now newly partially-running Vulkan backend  (which still has major accuracy issued with K_quants and only runs real _0 models). This Vulkan backend hangs at the end of Q8_0 generation, not openCL!

slaren commented:
Some results with Xiaomi 14 (snapdragon 8 gen 3):

| model                          |       size |     params | backend    | ngl |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------------: | -------------------: |
| llama 1B Q4_0                  | 727.75 MiB |     1.24 B | OpenCL     |  99 |         pp128 |        327.75 ± 1.58 |
| llama 1B Q4_0                  | 727.75 MiB |     1.24 B | OpenCL     |  99 |          tg32 |         23.10 ± 0.34 |
| llama 1B F16                   |   2.30 GiB |     1.24 B | OpenCL     |  99 |         pp128 |         33.68 ± 0.40 |
| llama 1B F16                   |   2.30 GiB |     1.24 B | OpenCL     |  99 |          tg32 |         18.78 ± 0.13 |

With `-ngl 0 -t 4` (CPU):
| model                          |       size |     params | backend    | ngl | threads |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ------------: | -------------------: |
| llama 1B Q4_0                  | 727.75 MiB |     1.24 B | OpenCL     |   0 |       4 |         pp128 |        259.01 ± 3.98 |
| llama 1B Q4_0                  | 727.75 MiB |     1.24 B | OpenCL     |   0 |       4 |          tg32 |         48.16 ± 0.47 |
| llama 1B F16                   |   2.30 GiB |     1.24 B | OpenCL     |   0 |       4 |         pp128 |         63.23 ± 0.31 |
| llama 1B F16                   |   2.30 GiB |     1.24 B | OpenCL     |   0 |       4 |          tg32 |         19.84 ± 1.45 |


<details>

```
ggml_opencl: selecting platform: 'QUALCOMM Snapdragon(TM)'
ggml_opencl: selecting device: 'QUALCOMM Adreno(TM) 750'
ggml_opencl: device OpenCL version: OpenCL 3.0 Adreno(TM) 750
ggml_opencl: OpenCL driver: OpenCL 3.0 QUALCOMM build: commit unknown Compiler E031.45.02.16
ggml_opencl: vector subgroup broadcast support: false
ggml_opencl: device FP16 support: true
ggml_opencl: mem base addr align: 1024
ggml_opencl: max mem alloc size: 1024 MB
ggml_opencl: SVM coarse grain buffer support: true
ggml_opencl: SVM fine grain buffer support: true
ggml_opencl: SVM fine grain system support: false
ggml_opencl: SVM atomics support: true
ggml_opencl: flattening quantized weights representation as struct of arrays (GGML_OPENCL_SOA_Q)
ggml_opencl: using kernels optimized for Adreno (GGML_OPENCL_USE_ADRENO_KERNELS)
```

</details>

sherylynn commented:
> > can u share how u build llama.cpp , my build show "ggml_opencl: plaform IDs not available. warning: no usable GPU found, --gpu-layers option will be ignored" As I use oneplus 13 8elite
> 
> The PR description (scroll to the top) includes instructions how to build with the Android NDK. I'm not familiar with Termux internals and how the env is setup. I typically just use NDK and ADB to build and run things.

build with the Android NDK and push to /data/local/tmp  can solve problem, thank you!!

myan-o commented:
> > > can u share how u build llama.cpp , my build show "ggml_opencl: plaform IDs not available. warning: no usable GPU found, --gpu-layers option will be ignored" As I use oneplus 13 8elite
> > 
> > The PR description (scroll to the top) includes instructions how to build with the Android NDK. I'm not familiar with Termux internals and how the env is setup. I typically just use NDK and ADB to build and run things.
> 
> build with the Android NDK and push to /data/local/tmp  can solve problem, thank you!!

The android ndk is also available in termux, so you can build it termux.
https://github.com/lzhiyong/termux-ndk

Francis235 commented:
使用Redmi K70(Qualcomm 8Gen 2)的一些结果：
./llama-bench-ocl -m Qwen2.5-0.5B-Instruct-Q4_0.gguf -t 4 -ngl 99
ggml_opencl: selecting platform: 'QUALCOMM Snapdragon(TM)'
ggml_opencl: selecting device: 'QUALCOMM Adreno(TM)'
ggml_opencl: OpenCL driver: OpenCL 3.0 QUALCOMM build: commit unknown Compiler E031.41.03.44
ggml_opencl: vector subgroup broadcast support: false
ggml_opencl: device FP16 support: true
ggml_opencl: mem base addr align: 128
ggml_opencl: max mem alloc size: 1024 MB
ggml_opencl: SVM coarse grain buffer support: true
ggml_opencl: SVM fine grain buffer support: true
ggml_opencl: SVM fine grain system support: false
ggml_opencl: SVM atomics support: true
ggml_opencl: flattening quantized weights representation as struct of arrays (GGML_OPENCL_SOA_Q)
ggml_opencl: using kernels optimized for Adreno (GGML_OPENCL_USE_ADRENO_KERNELS)
| model                          |       size |     params | backend    | ngl | threads |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ------------: | -------------------: |
| qwen2 1B Q4_0                  | 330.17 MiB |   494.03 M | OpenCL     |  99 |       4 |         pp512 |        212.38 ± 0.40 |
| qwen2 1B Q4_0                  | 330.17 MiB |   494.03 M | OpenCL     |  99 |       4 |         tg128 |         19.89 ± 0.12 |

./llama-bench-ocl -m Qwen2.5-0.5B-Instruct-Q4_0.gguf -t 4 -ngl 0
ggml_opencl: selecting platform: 'QUALCOMM Snapdragon(TM)'
ggml_opencl: selecting device: 'QUALCOMM Adreno(TM)'
ggml_opencl: OpenCL driver: OpenCL 3.0 QUALCOMM build: commit unknown Compiler E031.41.03.44
ggml_opencl: vector subgroup broadcast support: false
ggml_opencl: device FP16 support: true
ggml_opencl: mem base addr align: 128
ggml_opencl: max mem alloc size: 1024 MB
ggml_opencl: SVM coarse grain buffer support: true
ggml_opencl: SVM fine grain buffer support: true
ggml_opencl: SVM fine grain system support: false
ggml_opencl: SVM atomics support: true
ggml_opencl: flattening quantized weights representation as struct of arrays (GGML_OPENCL_SOA_Q)
ggml_opencl: using kernels optimized for Adreno (GGML_OPENCL_USE_ADRENO_KERNELS)
| model                          |       size |     params | backend    | ngl | threads |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ------------: | -------------------: |
| qwen2 1B Q4_0                  | 330.17 MiB |   494.03 M | OpenCL     |   0 |       4 |         pp512 |         75.51 ± 2.32 |
| qwen2 1B Q4_0                  | 330.17 MiB |   494.03 M | OpenCL     |   0 |       4 |         tg128 |         45.94 ± 1.15 |

I found that as ngl increases, the speed of prompt eval will become faster, but the speed of eval will become slower. I want to know if there is any way to increase the speed of prompt eval while maintaining the speed of eval? That is, use GPU in the prompt eval stage and use CPU in the eval stage to achieve the best generation speed.
